### PR TITLE
Thread safety for ml_tensors_info APIs

### PR DESCRIPTION
**Pull request by myungjoo**
Fixes #nnstreamer/2600

https://github.com/nnstreamer/nnstreamer/issues/2600

CC: @kparichay 
Signed-off-by: MyungJoo Ham <myungjoo.ham@samsung.com>

ps. This updates internal behaviors. No need for ACR.

### CODE REVIEW

File: c/src/nnstreamer-capi-util.c [hunk 1]
```diff
@@ -261,6 +294,8 @@ ml_tensors_info_set_count (ml_tensors_info_h info, unsigned int count)
     return ML_ERROR_INVALID_PARAMETER;
 
   tensors_info = (ml_tensors_info_s *) info;
+
+  /* This is atomic. No need for locks */
```
Comment (kparichay) [inline c/src/nnstreamer-capi-util.c:298]: I am not sure if this is an atomic operation, as this assignment can be broken down into more than 1 instruction by the compiler.

File: c/src/nnstreamer-capi-util.c [hunk 2]
```diff
@@ -198,6 +225,7 @@ ml_tensors_info_validate (const ml_tensors_info_h info, bool *valid)
   if (!tensors_info || tensors_info->num_tensors < 1)
```
Comment (kparichay) [inline c/src/nnstreamer-capi-util.c:225]: best to split this check into two parts and move the latter inside the lock.

**Review by kparichay — COMMENTED** [review]
Please add locks to `ml_tensors_info_clone()` as well, as it is an exposed API.
Also, for `ml_pipeline_custom_easy_filter_register()`, although each access to `ml_tensors_info_h` is locked, but are locked individually, and results in TOCTOU (tensors_info validity is checked and then later copied, but lock is released in between the checking and copying).

File: c/src/nnstreamer-capi-util.c (hunk 1 above)
Comment (myungjoo) [reply c/src/nnstreamer-capi-util.c:298]: 32bit (unsigned int) assignment cannot be more than 1 instruction. We may incur multiple instructions for redirections (```tensors_info->num_tensors```) and load (```LOAD Rx count```).
However, these does not affect the atomicity of assignment; they are allowed to be executed out of the critical section.
We cannot keep it serialized with destruction; thus, redirection is not protected anyway.
The location of ```count``` is never affected by the serialization problem; thus, atomicity is not affected.

Thus, we simply need to protect ```MOV Ra Rb``` instruction here.

File: c/src/nnstreamer-capi-util.c (hunk 2 above)
Comment (myungjoo) [reply c/src/nnstreamer-capi-util.c]: split done.

**Comment by myungjoo** [comment]
> Please add locks to `ml_tensors_info_clone()` as well, as it is an exposed API.
> Also, for `ml_pipeline_custom_easy_filter_register()`, although each access to `ml_tensors_info_h` is locked, but are locked individually, and results in TOCTOU (tensors_info validity is checked and then later copied, but lock is released in between the checking and copying).

added.

**Review by jaeyun-jung — APPROVED** [review]

**Comment by kparichay** [comment]
> > Please add locks to `ml_tensors_info_clone()` as well, as it is an exposed API.
> > Also, for `ml_pipeline_custom_easy_filter_register()`, although each access to `ml_tensors_info_h` is locked, but are locked individually, and results in TOCTOU (tensors_info validity is checked and then later copied, but lock is released in between the checking and copying).
> 
> added.

Can you please elaborate on how do you handle this? I can't see any `changes in ml_pipeline_custom_easy_filter_register()`.

**Comment by myungjoo** [comment]
> > > Please add locks to `ml_tensors_info_clone()` as well, as it is an exposed API.
> > > Also, for `ml_pipeline_custom_easy_filter_register()`, although each access to `ml_tensors_info_h` is locked, but are locked individually, and results in TOCTOU (tensors_info validity is checked and then later copied, but lock is released in between the checking and copying).
> > 
> > 
> > added.
> 
> Can you please elaborate on how do you handle this? I can't see any `changes in ml_pipeline_custom_easy_filter_register()`.

ah.. only clone was added. I'll add easy_filter soon.

**Comment by myungjoo** [comment]
> > > > Please add locks to `ml_tensors_info_clone()` as well, as it is an exposed API.
> > > > Also, for `ml_pipeline_custom_easy_filter_register()`, although each access to `ml_tensors_info_h` is locked, but are locked individually, and results in TOCTOU (tensors_info validity is checked and then later copied, but lock is released in between the checking and copying).
> > > 
> > > 
> > > added.
> > 
> > 
> > Can you please elaborate on how do you handle this? I can't see any `changes in ml_pipeline_custom_easy_filter_register()`.
> 
> ah.. only clone was added. I'll add easy_filter soon.

For ml_pipeline_custom_easy_filter_register(), it becomes safe if we don't access in or out after ml_tensors_info_clone; thus, I'll modify parameter usages of ml_tensors_info_copy_from_ml.

**Comment by kparichay** [comment]
> > > > > Please add locks to `ml_tensors_info_clone()` as well, as it is an exposed API.
> > > > > Also, for `ml_pipeline_custom_easy_filter_register()`, although each access to `ml_tensors_info_h` is locked, but are locked individually, and results in TOCTOU (tensors_info validity is checked and then later copied, but lock is released in between the checking and copying).
> > > > 
> > > > 
> > > > added.
> > > 
> > > 
> > > Can you please elaborate on how do you handle this? I can't see any `changes in ml_pipeline_custom_easy_filter_register()`.
> > 
> > 
> > ah.. only clone was added. I'll add easy_filter soon.
> 
> For ml_pipeline_custom_easy_filter_register(), it becomes safe if we don't access in or out after ml_tensors_info_clone; thus, I'll modify parameter usages of ml_tensors_info_copy_from_ml.

I think there is another issue of `tensors_info_h` being lock free between `ml_tensors_info_validate` and `ml_tensors_info_clone`.  As `ml_tensors_info_clone` does not check `ml_tensors_info_validate` on its own, it possible for a tensor to pass the validity check, but become non-valid before it could be cloned. Further, there is no check on result of the clone which I think is because we assume that all the tensors are valid, however that is not the case anymore because of the error of Time Of Check Time Of Use. I will make a separate PR to resolve this.

**Pull request merged**